File: echo-guardian-production-foundation-v1.0/native/ios/EchoGuardianKit/Sources/EchoGuardianKit/Sensors/MinimizedSensorObservation.swift
import Foundation
#if canImport(CoreMotion)
import CoreMotion
#endif

public enum SensorType: String, Codable {
    case accelerometer
    case gyroscope
    case devicePresence = "device_presence"
    case motion
}

public struct MinimizedSensorObservation: Codable, Equatable {
    public let schemaVersion: String
    public let observationId: String
    public let createdAt: String
    public let signalType: SensorType
    public let spaceId: String
    public let privateSpace: Bool
    public let permissionState: String
    public let availabilityState: String
    public let rawSensorRetained: Bool
    public let derivedMetadata: [String: Int]
    public let plainLanguageSummary: String

    public init(signalType: SensorType, spaceId: String, privateSpace: Bool, permissionState: String, availabilityState: String, derivedMetadata: [String: Int]) {
        precondition(privateSpace == false, "Private-space monitoring is blocked by default in v0.5")
        self.schemaVersion = "0.5"
        self.observationId = UUID().uuidString
        self.createdAt = ISO8601DateFormatter().string(from: Date())
        self.signalType = signalType
        self.spaceId = spaceId
        self.privateSpace = privateSpace
        self.permissionState = permissionState
        self.availabilityState = availabilityState
        self.rawSensorRetained = false
        self.derivedMetadata = derivedMetadata
        self.plainLanguageSummary = "Echo Guardian received a minimized safety signal for \(spaceId). Raw sensor data was not kept."
    }
}

public final class IOSSensorAdapter {
    #if canImport(CoreMotion)
    private let motionManager = CMMotionManager()
    #endif

    public init() {}

    public func accelerometerAvailability() -> String {
        #if canImport(CoreMotion)
        return motionManager.isAccelerometerAvailable ? "available" : "unavailable"
        #else
        return "unavailable_on_non_apple_runtime"
        #endif
    }

    public func makeMotionObservation(spaceId: String, confidenceScaled: Int, motionDetected: Bool) -> MinimizedSensorObservation {
        MinimizedSensorObservation(
            signalType: .motion,
            spaceId: spaceId,
            privateSpace: false,
            permissionState: "authorized_or_not_required",
            availabilityState: accelerometerAvailability(),
            derivedMetadata: [
                "confidence_scaled": max(0, min(100, confidenceScaled)),
                "motion_detected": motionDetected ? 1 : 0
            ]
        )
    }
}

File: echo-guardian-production-foundation-v1.0/native/android/echo-guardian-core/src/main/java/org/echoguardian/sensors/MinimizedSensorObservation.kt
package org.echoguardian.sensors

import android.content.Context
import android.hardware.Sensor
import android.hardware.SensorManager
import java.time.Instant
import java.util.UUID

data class MinimizedSensorObservation(
    val schemaVersion: String = "0.5",
    val observationId: String = UUID.randomUUID().toString(),
    val createdAt: String = Instant.now().toString(),
    val signalType: String,
    val spaceId: String,
    val privateSpace: Boolean,
    val permissionState: String,
    val availabilityState: String,
    val rawSensorRetained: Boolean = false,
    val derivedMetadata: Map<String, Int>,
    val plainLanguageSummary: String
)

class AndroidSensorAdapter(private val context: Context) {
    private val sensorManager: SensorManager = context.getSystemService(Context.SENSOR_SERVICE) as SensorManager

    fun accelerometerAvailability(): String =
        if (sensorManager.getDefaultSensor(Sensor.TYPE_ACCELEROMETER) != null) "available" else "unavailable"

    fun makeMotionObservation(spaceId: String, confidenceScaled: Int, motionDetected: Boolean): MinimizedSensorObservation {
        require(spaceId != "bathroom" && spaceId != "bedroom") { "Private-space monitoring is blocked by default in v0.5" }
        return MinimizedSensorObservation(
            signalType = "motion",
            spaceId = spaceId,
            privateSpace = false,
            permissionState = "authorized_or_not_required",
            availabilityState = accelerometerAvailability(),
            derivedMetadata = mapOf(
                "confidence_scaled" to confidenceScaled.coerceIn(0, 100),
                "motion_detected" to if (motionDetected) 1 else 0
            ),
            plainLanguageSummary = "Echo Guardian received a minimized safety signal for $spaceId. Raw sensor data was not kept."
        )
    }
}
